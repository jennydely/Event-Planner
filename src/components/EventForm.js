import React, { useRef } from 'react'
import { useForm } from 'react-hook-form'
import styled from 'styled-components/macro'
import Input from '../common/Input'
import Label from '../common/Label'
import ErrorMessage from '../common/ErrorMessage'
import Select from '../common/Select'
import PropTypes from 'prop-types'

EventForm.propTypes = {
  onEventSave: PropTypes.func.isRequired,
  packlists: PropTypes.array.isRequired,
}

export default function EventForm({ onEventSave, packlists }) {
  const { register, handleSubmit, errors, reset } = useForm()
  const onSubmit = (eventEntry, event) => {
    // for testing...
    if (event && event.target && typeof event.target.reset === 'function')
      event.target.reset()
    onEventSave(eventEntry)
  }
  const endDateRef = useRef(null)
  const allPacklists = packlists.map((packlist) => packlist.name)
  return (
    <>
      <Form data-testid="1337" onSubmit={handleSubmit(onSubmit)}>
        <CategoryInputLabel htmlFor="category">Category:</CategoryInputLabel>
        <CategoryInput
          name="category"
          id="category"
          register={register({ required: true })}
          options={['metal', 'medieval', 'sand', 'other']}
        />
        {errors.category && errors.category.type === 'required' && (
          <ErrorMessageCategoryReq>
            Category is required!
          </ErrorMessageCategoryReq>
        )}

        <InputLabelColumn2 row={2} htmlFor="name">
          EventName:
        </InputLabelColumn2>
        <InputColumn2
          row={3}
          placeholder="event name"
          id="name"
          name="name"
          ref={register({
            required: true,
            minLength: 3,
            maxLength: 20,
            validate: (value) =>
              value && value.trim().length >= 3 && value.trim().length <= 20,
          })}
        />

        {errors.name?.type === 'required' && (
          <ErrorMessageColumn2 row={4}>Name is required!</ErrorMessageColumn2>
        )}
        {(errors.name?.type === 'validate' ||
          errors.name?.type === 'minLength') && (
          <ErrorMessageColumn2 row={4}>
            This field requires at least 3 characters!
          </ErrorMessageColumn2>
        )}
        {(errors.name?.type === 'validate' ||
          errors.name?.type === 'maxLength') && (
          <ErrorMessageColumn2 row={4}>
            The name can reach a maximum of 20 characters!
          </ErrorMessageColumn2>
        )}

        <InputLabelColumn2 row={5} htmlFor="location">
          EventLocation:
        </InputLabelColumn2>
        <InputColumn2
          row={6}
          placeholder="location of the event"
          id="location"
          name="location"
          ref={register({
            required: true,
            minLength: 3,
            maxLength: 20,
            validate: (value) =>
              value && value.trim().length >= 3 && value.trim().length <= 20,
          })}
        />
        {errors.location?.type === 'required' && (
          <ErrorMessageColumn2 row={7}>
            Location is required!
          </ErrorMessageColumn2>
        )}
        {(errors.location?.type === 'validate' ||
          errors.location?.type === 'minLength') && (
          <ErrorMessageColumn2 row={7}>
            {' '}
            This field requires at least 3 characters!
          </ErrorMessageColumn2>
        )}
        {(errors.location?.type === 'validate' ||
          errors.location?.type === 'maxLength') && (
          <ErrorMessageColumn2 row={7}>
            The location can reach a maximum of 20 characters!
          </ErrorMessageColumn2>
        )}

        <EventInfosText>EventInfos</EventInfosText>
        <InputLabelColumn1 row={9} htmlFor="EventStartDate">
          Duration:
        </InputLabelColumn1>

        <EventStartDateInput
          placeholder="yyyy-mm-dd"
          id="EventStartDate"
          name="eventStartDate"
          ref={register({
            required: true,
            pattern: /([12]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01]))/,
            validate: (value) => value <= endDateRef.current.value,
          })}
        />
        {errors.eventStartDate?.type === 'validate' && (
          <ErrorMessageStartDate>
            Start date must be before end date!{' '}
          </ErrorMessageStartDate>
        )}

        {errors.eventStartDate?.type === 'required' && (
          <ErrorMessageStartDate>
            Date is required and needs the right form!{' '}
          </ErrorMessageStartDate>
        )}

        {errors.eventStartDate?.type === 'pattern' && (
          <ErrorMessageStartDate>
            Date must be written like: yyyy-mm-dd
          </ErrorMessageStartDate>
        )}

        <EventEndDateInput
          placeholder="yyyy-mm-dd"
          id="EventEndDate"
          name="eventEndDate"
          ref={(el) => {
            endDateRef.current = el
            register(el, {
              required: true,
              pattern: /([12]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01]))/,
            })
          }}
        />

        {errors.eventEndDate?.type === 'required' && (
          <ErrorMessageEndDate>
            Date is required and needs the right form!{' '}
          </ErrorMessageEndDate>
        )}

        {errors.eventEndDate?.type === 'pattern' && (
          <ErrorMessageEndDate>
            {' '}
            Date must be written like: yyyy-mm-dd{' '}
          </ErrorMessageEndDate>
        )}

        <AddressInputLabel htmlFor="Street">Address:</AddressInputLabel>
        <InputColumn2
          row={11}
          placeholder="street + number"
          id="street"
          name="street"
          ref={register({
            required: true,
            minLength: 5,
            validate: (value) => value && value.trim().length >= 5,
          })}
        />

        {errors.street?.type === 'required' && (
          <ErrorMessageColumn2 row={12}>
            Street is required!
          </ErrorMessageColumn2>
        )}
        {(errors.street?.type === 'validate' ||
          errors.street?.type === 'minLength') && (
          <ErrorMessageColumn2 row={12}>
            This field requires at least 5 characters!
          </ErrorMessageColumn2>
        )}

        <InputColumn2
          row={13}
          placeholder="zip"
          id="zip"
          name="zip"
          ref={register({
            required: true,
            minLength: 2,
            validate: (value) => value && value.trim().length >= 2,
          })}
        />

        {errors.zip?.type === 'required' && (
          <ErrorMessageColumn2 row={14}>Zip is required!</ErrorMessageColumn2>
        )}
        {(errors.zip?.type === 'validate' ||
          errors.zip?.type === 'minLength') && (
          <ErrorMessageColumn2 row={14}>
            This field requires at least 2 characters!
          </ErrorMessageColumn2>
        )}

        <InputLabelColumn1 row={15} htmlFor="website">
          Website:
        </InputLabelColumn1>
        <InputColumn2
          row={15}
          placeholder="http://www.website.de"
          id="website"
          name="website"
          ref={register({
            required: true,
            minLength: 8,
            validate: (value) => value && value.trim().length >= 8,
            pattern: /(https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|www\.[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9]+\.[^\s]{2,}|www\.[a-zA-Z0-9]+\.[^\s]{2,})/,
          })}
        />

        {errors.website?.type === 'required' && (
          <ErrorMessageColumn2 row={16}>
            Website is required!
          </ErrorMessageColumn2>
        )}
        {(errors.website?.type === 'validate' ||
          errors.website?.type === 'minLength') && (
          <ErrorMessageColumn2 row={16}>
            This field requires at least 8 characters!
          </ErrorMessageColumn2>
        )}
        {errors.website?.type === 'pattern' && (
          <ErrorMessageColumn2 row={16}>
            Should begin with www or http
          </ErrorMessageColumn2>
        )}

        <InputLabelColumn1 row={17} htmlFor="price">
          TicketPrice:
        </InputLabelColumn1>
        <InputColumn2
          row={17}
          placeholder="ticket price or range"
          id="price"
          name="price"
          ref={register({ required: false })}
        />
        <InputLabelColumn1 row={18} htmlFor="poster">
          Picture:{' '}
        </InputLabelColumn1>
        <InputColumn2
          row={18}
          placeholder="http://www.website.de/banner.jpg"
          id="poster"
          name="poster"
          ref={register({
            required: false,
            pattern: /(https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|www\.[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9]+\.[^\s]{2,}|www\.[a-zA-Z0-9]+\.[^\s]{2,})/,
          })}
        />
        <InputLabelColumn1 row={19} htmlFor="packlist">
          Add PackList:
        </InputLabelColumn1>
        <PacklistInput
          name="packlistCategory"
          id="packlist"
          register={register({ required: true })}
          options={allPacklists}
        />
        <ButtonGroup>
          <button type="reset" onClick={() => reset()}>
            Reset
          </button>
          <SubmitButton type="submit">Save</SubmitButton>
        </ButtonGroup>
      </Form>
    </>
  )
}

const Form = styled.form`
  display: grid;
  grid-template-columns: repeat(3, auto);
  grid-template-rows: repeat(20, auto);
  align-content: center;
  min-width: 300px;
  gap: 4px;
`
const InputLabelColumn1 = styled(Label)`
  grid-column: 1;
  grid-row: ${(props) => props.row};
`
const InputLabelColumn2 = styled(Label)`
  grid-column: 2 / span 2;
  grid-row: ${(props) => props.row};
  text-align: left;
`
const InputColumn2 = styled(Input)`
  grid-column: 2 / span 2;
  grid-row: ${(props) => props.row};
`
const ErrorMessageColumn2 = styled(ErrorMessage)`
  grid-column: 2 / span 2;
  grid-row: ${(props) => props.row};
`
const CategoryInputLabel = styled(Label)`
  grid-column: 1;
  grid-row: 2;
  text-align: left;
`
const CategoryInput = styled(Select)`
  grid-column: 1;
  grid-row: 3;
  display: block;
  width: 100%;
  padding: 20px;
  border: 1px solid rgb(96, 99, 104);
  border-radius: 4px;
  margin-top: 0;
  padding: 4px;
  font-size: 112.5%;
  color: black;
  background: ${({ name }) =>
    name === 'sand'
      ? 'rgba(248,149,17,0.46)'
      : name === 'metal'
      ? 'rgba(49,42,42,0.75)'
      : name === 'medieval'
      ? 'rgba(67,40,24,0.70)'
      : name === 'other'
      ? 'rgba(153,88,42,0.70)'
      : 'rgb(96,99,104)'};
`
const ErrorMessageCategoryReq = styled(ErrorMessage)`
  grid-column: 1;
  grid-row: 4;
`
const EventInfosText = styled.h2`
  grid-column: 1 / span 2;
  grid-row: 8;
  margin-top: 20px;
`
const EventStartDateInput = styled(Input)`
  grid-column: 2;
  grid-row: 9;
`
const ErrorMessageStartDate = styled(ErrorMessage)`
  grid-column: 2;
  grid-row: 10;
`
const EventEndDateInput = styled(Input)`
  grid-column: 3;
  grid-row: 9;
`
const ErrorMessageEndDate = styled(ErrorMessage)`
  grid-column: 3;
  grid-row: 10;
`
const AddressInputLabel = styled(Label)`
  grid-column: 1;
  grid-row: 11 / span 2;
`
const PacklistInput = styled(Select)`
  grid-column: 2 / span 2;
  grid-row: 19;
  display: block;
  width: 100%;
  border-radius: 4px;
  border: 1px solid rgb(96, 99, 104);
  margin-top: 0;
  padding: 4px;
  font-size: 112.5%;
  color: black;
  background: rgb(96, 99, 104);
`
const ButtonGroup = styled.div`
  grid-column: 1 / span 3;
  grid-row: 20;
  display: flex;
  justify-content: center;
  width: 100%;
  margin: 20px 7px;
`
const SubmitButton = styled.button`
  grid-column: 1 / span 3;
  grid-row: 20;
  background-color: rgba(111, 29, 27, 0.75);
`
