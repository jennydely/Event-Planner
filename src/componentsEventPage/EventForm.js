import React, { useRef } from 'react'
import { useForm } from 'react-hook-form'
import styled from 'styled-components/macro'
import Input from '../common/Input'
import Label from '../common/Label'
import ErrorMessage from '../common/ErrorMessage'
import Select from '../common/Select'
import PropTypes from 'prop-types'

EventForm.propTypes = {
    onSave: PropTypes.func.isRequired,
}

export default function EventForm({ onSave }) {
    const { register, handleSubmit, errors, reset } = useForm()
    const onSubmit = (eventEntry, event) => {
        // for testing...
        if (event && event.target && typeof event.target.reset === 'function')
            event.target.reset()
        onSave(eventEntry)
    }
    const endDateRef = useRef(null)

    return (
        <>
            <Form data-testid="1337" onSubmit={handleSubmit(onSubmit)}>
                <TitleText>Create Event</TitleText>
                <CategoryInputLabel htmlFor="category">
                    Category:
                    </CategoryInputLabel>
                <CategoryInput name="category" id="category"
                    register={register({ required: true })} options={["metal", "medieval", "sand", "other"]} />
                {errors.category && errors.category.type === 'required' && (
                    <ErrorMessageCategoryReq>Category is required!</ErrorMessageCategoryReq>
                )}

                <NameInputLabel htmlFor="name">
                    EventName:
                    </NameInputLabel>
                <NameInput
                    placeholder="event name"
                    id="name"
                    name="name"
                    ref={register({
                        required: true, minLength: 3, maxLength: 20,
                        validate: value => value && value.trim().length >= 3 && value.trim().length <= 20
                    })}
                />

                {errors.name && errors.name.type === 'required' && (
                    <ErrorMessageNameReq>Name is required!</ErrorMessageNameReq>
                )}
                {errors.name && (errors.name.type === 'validate' || errors.name.type === 'minLength') && (
                    <ErrorMessageName>
                        This field requires at least 3 characters!
                    </ErrorMessageName>
                )}
                {errors.name && (errors.name.type === 'validate' || errors.name.type === 'maxLength') && (
                    <ErrorMessageName>
                        The name can reach a maximum of 20 characters!
                    </ErrorMessageName>
                )}


                <LocationInputLabel htmlFor="location">
                    EventLocation:
                    </LocationInputLabel>
                <LocationInput
                    placeholder="location of the event"
                    id="location"
                    name="location"
                    ref={register({
                        required: true,
                        minLength: 3,
                        maxLength: 20,
                        validate: value => value && value.trim().length >= 3 && value.trim().length <= 20
                    })}
                />

                {errors.location && errors.location.type === 'required' && (
                    <ErrorMessageLocationReq>Location is required!</ErrorMessageLocationReq>
                )}
                {errors.location && (errors.location.type === 'validate' || errors.location.type === 'minLength') && (
                    <ErrorMessageLocation> This field requires at least 3 characters!</ErrorMessageLocation>
                )}
                {errors.location && (errors.location.type === 'validate' || errors.location.type === 'maxLength') && (
                    <ErrorMessageLocation>
                        The location can reach a maximum of 20 characters!
                    </ErrorMessageLocation>
                )}

                <EventInfosText>EventInfos</EventInfosText>
                <DurationInputLabel htmlFor="EventStartDate">
                    Duration:</DurationInputLabel>

                <EventStartDateInput
                    placeholder="yyyy-mm-dd"
                    id="EventStartDate"
                    name="eventStartDate"
                    ref={register({
                        required: true,
                        pattern: /([12]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01]))/,
                        validate: value => value <= endDateRef.current.value,
                    })}
                />
                {(errors.eventStartDate && errors.eventStartDate.type === 'validate') &&
                    <ErrorMessageStartDate>Start date must be before end date! </ErrorMessageStartDate>
                }

                {(errors.eventStartDate && errors.eventStartDate.type === 'required') &&
                    <ErrorMessageStartDate>Date is required and needs the right form! </ErrorMessageStartDate>
                }

                {(errors.eventStartDate && errors.eventStartDate.type === 'pattern') &&
                    (<ErrorMessageStartDate>Date must be written like: yyyy-mm-dd</ErrorMessageStartDate>)}

                <EventEndDateInput
                    placeholder="yyyy-mm-dd"
                    id="EventEndDate"
                    name="eventEndDate"
                    ref={(el) => {
                        endDateRef.current = el
                        register(el, {
                            required: true,
                            pattern: /([12]\d{3}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01]))/,
                        })
                    }}
                />

                {(errors.eventEndDate && errors.eventEndDate.type === 'required') &&
                    <ErrorMessageEndDate>Date is required and needs the right form! </ErrorMessageEndDate>
                }

                {(errors.eventEndDate && errors.eventEndDate.type === 'pattern') &&
                    (
                        <ErrorMessageEndDate> Date must be written like: yyyy-mm-dd </ErrorMessageEndDate>
                    )}

                <AddressInputLabel htmlFor="Street">
                    Address:
                </AddressInputLabel>
                <StreetInput
                    placeholder="street + number"
                    id="street"
                    name="street"
                    ref={register({
                        required: true, minLength: 5,
                        validate: value => value && value.trim().length >= 5
                    })}
                />

                {errors.street && errors.street.type === 'required' && (
                    <ErrorMessageStreetReq>Street is required!</ErrorMessageStreetReq>
                )}
                {errors.street && (errors.street.type === 'validate' || errors.street.type === 'minLength') && (
                    <ErrorMessageStreet>
                        This field requires at least 5 characters!
                    </ErrorMessageStreet>
                )}

                <ZipInput
                    placeholder="zip"
                    id="zip"
                    name="zip"
                    ref={register({
                        required: true, minLength: 2,
                        validate: value => value && value.trim().length >= 2
                    })}
                />

                {errors.zip && errors.zip.type === 'required' && (
                    <ErrorMessageZipReq>Zip is required!</ErrorMessageZipReq>
                )}
                {errors.zip && (errors.zip.type === 'validate' || errors.zip.type === 'minLength') && (
                    <ErrorMessageZip>
                        This field requires at least 2 characters!
                    </ErrorMessageZip>
                )}

                <WebsiteInputLabel htmlFor="website">
                    Website:</WebsiteInputLabel>
                <WebsiteInput
                    placeholder="http://www.website.de"
                    id="website"
                    name="website"
                    ref={register({
                        required: true,
                        minLength: 8,
                        validate: value => value && value.trim().length >= 8,
                        pattern: /(https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|www\.[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9]+\.[^\s]{2,}|www\.[a-zA-Z0-9]+\.[^\s]{2,})/,
                    })}
                />

                {errors.website && errors.website.type === 'required' && (
                    <ErrorMessageWebsiteReq>Website is required!</ErrorMessageWebsiteReq>
                )}
                {errors.website && (errors.website.type === 'validate' || errors.website.type === 'minLength') && (
                    <ErrorMessageWebsite>
                        This field requires at least 8 characters!
                    </ErrorMessageWebsite>
                )}

                <PriceInputLabel htmlFor="price">
                    TicketPrice:</PriceInputLabel>
                <PriceInput
                    placeholder="ticket price or range"
                    id="price"
                    name="price"
                    ref={register({ required: false })}
                />

                <PictureInputLabel htmlFor="poster">
                    Picture: </PictureInputLabel>
                <PictureInput
                    placeholder="http://www.website.de/banner.jpg"
                    id="poster"
                    name="poster"
                    ref={register({
                        required: false,
                        pattern: /(https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|www\.[a-zA-Z0-9][a-zA-Z0-9-]+[a-zA-Z0-9]\.[^\s]{2,}|https?:\/\/(?:www\.|(?!www))[a-zA-Z0-9]+\.[^\s]{2,}|www\.[a-zA-Z0-9]+\.[^\s]{2,})/,
                    })}
                />

                <PacklistInputLabel htmlFor="packlist">
                    Add PackList:
                    </PacklistInputLabel>
                <PacklistInput name="packlistCategory" id="packlist"
                    register={register({ required: true })} options={["festival", "medieval"]} />

                <ButtonGroup>
                    <button type="reset" onClick={() => reset()}>
                        Reset
                    </button>
                    <SubmitButton type="submit">Save</SubmitButton>
                </ButtonGroup>
            </Form>
        </>
    )
}

const Form = styled.form`
display:grid;
grid-template-columns: repeat(3,auto);
grid-template-rows: repeat(19,auto) 30px;
align-content: center;
min-width: 300px;
gap: 4px;
`
const TitleText = styled.h1`
grid-column: 1 / span 3;
grid-row: 1;
`
const CategoryInputLabel = styled(Label)`
grid-column: 1;
grid-row: 2;
text-align:left;
`
const CategoryInput = styled(Select)`
grid-column: 1;
grid-row: 3;
display: block;
width: 100%;
padding: 20px;
border: 1px solid rgb(96,99,104);
border-radius: 4px;
margin-top: 0;
padding: 4px;
font-size: 100%;
color: black;
background: ${({ name }) => name === 'sand' ? 'rgba(248,149,17,0.46)' : (name === 'metal' ? 'rgba(49,42,42,0.75)' : (name === 'medieval' ? 'rgba(67,40,24,0.70)' : (name === 'other' ? 'rgba(153,88,42,0.70)' : 'rgb(96,99,104)')))};
`
const ErrorMessageCategoryReq = styled(ErrorMessage)`
grid-column: 1;
grid-row: 4;
`
const NameInputLabel = styled(Label)`
grid-column: 2 / span 2;
grid-row: 2;
text-align:left;
`
const NameInput = styled(Input)`
grid-column: 2 / span 2;
grid-row: 3;
`
const ErrorMessageNameReq = styled(ErrorMessage)`
grid-column: 2 / span 2;
grid-row: 4;
`
const ErrorMessageName = styled(ErrorMessage)`
grid-column: 2 / span 2;
grid-row: 4;
`
const LocationInputLabel = styled(Label)`
grid-column: 2 / span 2;
grid-row: 5;
text-align:left;
`
const LocationInput = styled(Input)`
grid-column: 2 / span 2;
grid-row: 6;
`
const ErrorMessageLocationReq = styled(ErrorMessage)`
grid-column: 2 / span 2;
grid-row: 7;
`
const ErrorMessageLocation = styled(ErrorMessage)`
grid-column: 2 / span 2;
grid-row: 7;
`
const EventInfosText = styled.h2`
grid-column: 1/ span 2;
grid-row: 8;
margin-top:30px;
`
const DurationInputLabel = styled(Label)`
grid-column: 1;
grid-row: 9;
`
const EventStartDateInput = styled(Input)`
grid-column: 2;
grid-row: 9;
`
const ErrorMessageStartDate = styled(ErrorMessage)`
grid-column: 2;
grid-row: 10;
`
const EventEndDateInput = styled(Input)`
grid-column: 3;
grid-row: 9;
`
const ErrorMessageEndDate = styled(ErrorMessage)`
grid-column: 3;
grid-row: 10;
`
const AddressInputLabel = styled(Label)`
grid-column: 1;
grid-row: 11 /span 2 ;
`
const StreetInput = styled(Input)`
grid-column: 2 / span 2;
grid-row: 11 ;
`
const ErrorMessageStreetReq = styled(ErrorMessage)`
grid-column: 2 / span 2;
grid-row: 12 ;
`
const ErrorMessageStreet = styled(ErrorMessage)`
grid-column: 2 / span 2;
grid-row: 12 ;
`
const ZipInput = styled(Input)`
grid-column: 2 / span 2;
grid-row: 13 ;
`
const ErrorMessageZipReq = styled(ErrorMessage)`
grid-column: 2 / span 2;
grid-row: 14 ;
`
const ErrorMessageZip = styled(ErrorMessage)`
grid-column: 2 / span 2;
grid-row: 14 ;
`
const WebsiteInputLabel = styled(Label)`
grid-column: 1;
grid-row: 15;
`
const WebsiteInput = styled(Input)`
grid-column: 2 / span 2;
grid-row: 15;
`
const ErrorMessageWebsiteReq = styled(ErrorMessage)`
grid-column: 2 / span 2;
grid-row: 16;
`
const ErrorMessageWebsite = styled(ErrorMessage)`
grid-column: 2 / span 2;
grid-row: 16;
`
const PriceInputLabel = styled(Label)`
grid-column: 1;
grid-row: 17;
`
const PriceInput = styled(Input)`
grid-column: 2 / span 2;
grid-row: 17;
`
const PictureInputLabel = styled(Label)`
grid-column: 1;
grid-row: 18;
`
const PictureInput = styled(Input)`
grid-column: 2 / span 2;
grid-row: 18;
`

const PacklistInputLabel = styled(Label)`
grid-column: 1;
grid-row: 19;
`

const PacklistInput = styled(Select)`
grid-column: 2 / span 2;
grid-row: 19;
display: block;
width: 100%;
padding: 20px;
border-radius: 4px;
border: 1px solid rgb(96,99,104);
margin-top: 0;
padding: 4px;
font-size: 100%;
color: black;
background: rgb(96,99,104);
`

const ButtonGroup = styled.div`
grid-column: 1 / span 3;
grid-row: 20;
display: flex;
justify-content: center;
width: 100%;
maring:7px;
margin-top: 30px;
`
const SubmitButton = styled.button`
  background-color: rgba(111,29,27,0.75);
`

