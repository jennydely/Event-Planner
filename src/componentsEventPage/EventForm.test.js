import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react'
import renderer from 'react-test-renderer'
import EventForm from './EventForm'
import 'jest-styled-components'

window.MutationObserver = require("mutation-observer");

describe('EventForm', () => {

  it('creates a new event', async () => {
    const onSave = jest.fn()
    const { getByText, getByLabelText, getByPlaceholderText, getAllByPlaceholderText } = render(<EventForm onSave={onSave} />)

    fireEvent.change(getByLabelText('Category:'), { target: { value: 'medieval' } })
    fireEvent.input(getByPlaceholderText('event name'), { target: { value: 'My EventName' } })
    fireEvent.input(getByPlaceholderText('location of the event'), { target: { value: 'My EventLocation' } })
    fireEvent.input(getByPlaceholderText('street + number'), { target: { value: 'My Address' } })
    fireEvent.input(getByPlaceholderText('http://www.website.de'), { target: { value: 'www.website.de' } })
    fireEvent.input(getByPlaceholderText('zip'), { target: { value: '12345' } })
    fireEvent.input(getAllByPlaceholderText('yyyy-mm-dd')[0], { target: { value: '2020-05-20' } })
    fireEvent.input(getAllByPlaceholderText('yyyy-mm-dd')[1], { target: { value: '2020-05-25' } })

    fireEvent.submit(getByText('Save'))
    await waitFor(() => expect(onSave).toHaveBeenCalled())
  })

  it('displays a form to create events', async () => {
    const onSave = jest.fn()
    const { getByText, getByLabelText, getByPlaceholderText } = render(<EventForm onSave={onSave} />)

    const nameInput = getByPlaceholderText('event name')
    expect(nameInput.value).toBe('')
    fireEvent.input(nameInput, { target: { value: 'My EventName' } })
    expect(nameInput.value).toBe('My EventName')

    fireEvent.click(getByText('Reset'))
    await waitFor(() => expect(nameInput.value).toBe(''))
    expect(nameInput.value).toBe('')
    await waitFor(() => expect(getByLabelText('Category:').value).toBe('metal'))
    expect(getByLabelText('Category:').value).toBe('metal')
  /*expect(getByLabel('name').toBeInTheDocument()

  expect(getByLabel('location').toBeInTheDocument()

  expect(getByText('Duration:')).toBeInTheDocument()

  expect(getByText('Address:')).toBeInTheDocument()

  expect(getAllByRole('button')).toBeInTheDocument()
  
  await fireEvent.submit(getByTestId('1337')). */
})

it('renders correctly', () => {
  const onSave = jest.fn()
  const tree = renderer.create(<EventForm onSave={onSave} />)
  expect(tree).toMatchSnapshot()
})
})
