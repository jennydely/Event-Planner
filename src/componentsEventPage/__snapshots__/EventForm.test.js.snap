// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventForm renders correctly 1`] = `
.c6 {
  display: block;
  width: 100%;
  padding: 20px;
  border-radius: 4px;
  border: 1px solid var(--blue-50);
  margin-top: 0;
  padding: 4px;
  font-size: 100%;
  color: black;
}

.c2 {
  color: var(--blue-main);
  font-weight: 400;
  font-size: 100%;
  line-height: 1.143;
  margin: 4px;
  text-align: right;
}

.c0 {
  display: grid;
  grid-template-columns: repeat(3,auto);
  grid-template-rows: repeat(18,auto) 30px;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  min-width: 300px;
  gap: 4px;
}

.c1 {
  grid-column: 1 / span 3;
  grid-row: 1;
}

.c3 {
  grid-column: 1;
  grid-row: 2;
  text-align: left;
}

.c4 {
  grid-column: 1;
  grid-row: 3;
  width: 100%;
  padding: 20px;
  border-radius: 4px;
  border: 1px solid var(--blue-50);
  margin-top: 0;
  padding: 4px;
  font-size: 100%;
  color: black;
}

.c5 {
  grid-column: 2 / span 2;
  grid-row: 2;
  text-align: left;
}

.c7 {
  grid-column: 2 / span 2;
  grid-row: 3;
}

.c8 {
  grid-column: 2 / span 2;
  grid-row: 5;
  text-align: left;
}

.c9 {
  grid-column: 2 / span 2;
  grid-row: 6;
}

.c10 {
  grid-column: 1/ span 2;
  grid-row: 8;
  margin-top: 30px;
}

.c11 {
  grid-column: 1;
  grid-row: 9;
}

.c12 {
  grid-column: 2;
  grid-row: 9;
}

.c13 {
  grid-column: 3;
  grid-row: 9;
}

.c14 {
  grid-column: 1;
  grid-row: 11 /span 2;
}

.c15 {
  grid-column: 2 / span 2;
  grid-row: 11;
}

.c16 {
  grid-column: 2 / span 2;
  grid-row: 13;
}

.c17 {
  grid-column: 1;
  grid-row: 15;
}

.c18 {
  grid-column: 2 / span 2;
  grid-row: 15;
}

.c19 {
  grid-column: 1;
  grid-row: 17;
}

.c20 {
  grid-column: 2 / span 2;
  grid-row: 17;
}

.c21 {
  grid-column: 1;
  grid-row: 18;
}

.c22 {
  grid-column: 2 / span 2;
  grid-row: 18;
}

.c23 {
  grid-column: 1 / span 3;
  grid-row: 19;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  maring: 7px;
  margin-top: 30px;
}

.c24 {
  background-color: rgba(111,29,27,0.75);
}

<form
  className="c0"
  data-testid="1337"
  onSubmit={[Function]}
>
  <h1
    className="c1"
  >
    Create Event
  </h1>
  <label
    className="c2 c3"
    htmlFor="category"
  >
    Category:
  </label>
  <select
    className="c4"
    id="category"
    name="category"
  >
    <option
      value="metal"
    >
      metal
    </option>
    <option
      value="medieval"
    >
      medieval
    </option>
    <option
      value="sand"
    >
      sand
    </option>
    <option
      value="other"
    >
      other
    </option>
  </select>
  <label
    className="c2 c5"
    htmlFor="name"
  >
    EventName:
  </label>
  <input
    className="c6 c7"
    id="name"
    name="name"
    placeholder="event name"
  />
  <label
    className="c2 c8"
    htmlFor="location"
  >
    EventLocation:
  </label>
  <input
    className="c6 c9"
    id="location"
    name="location"
    placeholder="location of the event"
  />
  <h2
    className="c10"
  >
    EventInfos
  </h2>
  <label
    className="c2 c11"
    htmlFor="EventStartDate"
  >
    Duration:
  </label>
  <input
    className="c6 c12"
    id="EventStartDate"
    name="eventStartDate"
    placeholder="yyyy-mm-dd"
  />
  <input
    className="c6 c13"
    id="EventEndDate"
    name="eventEndDate"
    placeholder="yyyy-mm-dd"
  />
  <label
    className="c2 c14"
    htmlFor="Street"
  >
    Address:
  </label>
  <input
    className="c6 c15"
    id="street"
    name="street"
    placeholder="street + number"
  />
  <input
    className="c6 c16"
    id="zip"
    name="zip"
    placeholder="zip"
  />
  <label
    className="c2 c17"
    htmlFor="website"
  >
    Website:
  </label>
  <input
    className="c6 c18"
    id="website"
    name="website"
    placeholder="http://www.website.de"
  />
  <label
    className="c2 c19"
    htmlFor="price"
  >
    TicketPrice:
  </label>
  <input
    className="c6 c20"
    id="price"
    name="price"
    placeholder="ticket price or range"
  />
  <label
    className="c2 c21"
    htmlFor="poster"
  >
    Picture: 
  </label>
  <input
    className="c6 c22"
    id="poster"
    name="poster"
    placeholder="http://www.website.de/banner.jpg"
  />
  <div
    className="c23"
  >
    <button
      onClick={[Function]}
      type="reset"
    >
      Reset
    </button>
    <button
      className="c24"
      type="submit"
    >
      Save
    </button>
  </div>
</form>
`;
